dynamic/run.sh

```diff
@@ -94,51 +94,132 @@ rm -rf /tmp/* /opt/instana/agent/etc/org.ops4j.pax.logging.cfg \
+        cat /tmp/ecs_data
+        exit 1
+      fi
+      ;;
 
-    ROLES_FOUND=false
+    'EC2')
+      readonly EC2_METADATA_ENDPOINT='http://169.254.169.254/latest/dynamic/instance-identity/document'
 
-    if ! curl -s http://169.254.169.254/latest/meta-data/iam/security-credentials/ --connect-timeout 2 | grep 404&> /dev/null; then
-      ROLES_FOUND=true
-    fi
+      if ! curl -s "${EC2_METADATA_ENDPOINT}" --connect-timeout 2 --fail -o /tmp/ec2_data; then
+        echo "Cannot retrieve metadata from the '${EC2_METADATA_ENDPOINT}' endpoint. Aborting startup."
+        exit 1
+      fi
+
+      if ! INSTANA_AWS_REGION_CONFIG=$(jq -e -r '.region' < /tmp/ec2_data); then
+        echo "The metadata endpoint did not return the 'region' key. Aborting startup."
+        cat /tmp/ecs_data
+        exit 1
+      fi
+      ;;
 
-    if [ "$ROLES_FOUND" = "false" ]; then
-      if [ -z "$AWS_ACCESS_KEY_ID" ] || [ -z "$AWS_SECRET_ACCESS_KEY" ]; then
-        echo "AWS_ACCESS_KEY_ID and/or AWS_SECRET_ACCESS_KEY not exported, and no IAM instance role detected to allow AWS API access."
+    *)
+      if [ -n "${AWS_DEFAULT_REGION}" ]; then
+        echo "Using the default AWS region '${AWS_DEFAULT_REGION}' set in the environment via the 'AWS_DEFAULT_REGION' environment variable."
+        INSTANA_AWS_REGION_CONFIG="${AWS_DEFAULT_REGION}"
+      else
+        echo "Platform not recognized: this agent does not seem to run on EC2 or ECS. Set the 'INSTANA_AWS_REGION_CONFIG' environment variable. Aborting startup."
         exit 1
       fi
+      ;;
+    esac
+
+    if [ -z "${INSTANA_AWS_REGION_CONFIG}" ]; then
+      echo "Could not retrieve the AWS region from the AWS metadata. Aborting startup."
+      exit 1
     fi
 
-    echo -e "\nmode = INFRASTRUCTURE" >> /opt/instana/agent/etc/instana/com.instana.agent.main.config.Agent.cfg
-  else
-    echo -e "\nmode = ${INSTANA_AGENT_MODE}" >> /opt/instana/agent/etc/instana/com.instana.agent.main.config.Agent.cfg
+    echo "Discovered AWS region: ${INSTANA_AWS_REGION_CONFIG}"
+    export INSTANA_AWS_REGION_CONFIG
+  fi
+
+  if [ "${PLATFORM}" = 'UNKNOWN' ]; then
+    if [ -z "${AWS_ACCESS_KEY_ID}" ] || [ -z "${AWS_SECRET_ACCESS_KEY}" ]; then
+      echo "The platform is not recognized (this agent does not seem to run on EC2 or ECS) and neither 'AWS_ACCESS_KEY_ID' nor 'AWS_SECRET_ACCESS_KEY' environment variables are exported. AWS Services monitoring might just not work. If so, please configure either the 'AWS_ACCESS_KEY_ID' or the 'AWS_SECRET_ACCESS_KEY' environment variables."
+    fi
   fi
 fi
 
+# Normalize the Agent mode to those that are actually used by the agent itself, i.e., OFF, APM and INFRASTRUCTURE
+case "${INSTANA_AGENT_MODE}" in
+  'AWS')
+    INSTANA_AGENT_MODE='INFRASTRUCTURE'
+    ;; # The AWS agent mode is infra + AWS sensor setup
+  'INFRASTRUCTURE')
+    ;;
+  'APM')
+    ;;
+  'OFF')
+    ;;
```

Comment:
Doesn't matter much, but if `case` entries are the same you can also use the pipe `|` to combine them. And quoting is not needed. So
`INFRASTRUCTURE|APM|OFF)`
would be the same.

Reply:
But from readability would probably also leave it like this :-)
But can be handy when covering multiple case-sensitive options like `y|Y|yes|YES|Yes)` for example.